### PR TITLE
GODRIVER-2154 Audit code generated by operationgen and remove operationgen

### PR DESCRIPTION
**Pull request by benjirewis**
GODRIVER-2154

The Go driver used to rely on an internal code generation tool called [`operationgen`](https://github.com/mongodb/mongo-go-driver/tree/master/cmd/operationgen) to create the operation-specific files under `x/mongo/driver/operation`. We no longer rely on this tool and now edit these files by hand.

This PR does a number of things:
1. Audits all `build[type]Result` functions in operations under `x/mongo/driver/operation` and correctly returns errors (fixes a number of bugs created by a faulty `operationgen` template).
2. Removes all of the `// Code generated by operationgen. DO NOT EDIT.` notes at the top of files in `x/mongo/driver/operation`.
3. Removes all `*.toml` files in `x/mongo/driver/operation`.
4. Removes `cmd/operationgen` and `x/mongo/driver/drivergen`.
5. Re-enables "staticcheck" and "ineffassign" linters for `x/mongo/driver/operation` in `.golangci.yml`.

The `operationgen` also used a dependency called [`packr`](https://github.com/gobuffalo/packr) that I believe we can now remove. To avoid bloating this PR, I've filed GODRIVER-2234 to do that separately.

### CODE REVIEW

File: x/mongo/driver/operation/find_and_modify.go [hunk 1]
```diff
@@ -74,23 +72,23 @@ func buildFindAndModifyResult(response bsoncore.Document, srvr driver.Server) (F
...
 	for _, element := range elements {
 		switch element.Key() {
-
 		case "value":
 			var ok bool
 			famr.Value, ok = element.Value().DocumentOK()
-			if !ok {
-				err = fmt.Errorf("response field 'value' is type document, but received BSON type %s", element.Value().Type)
+
+			// The 'value' field returned by a FindAndModify can be null in the case that no document was found.
```
Comment (benjirewis) [inline x/mongo/driver/operation/find_and_modify.go:79]: Calling attention to this bug-fix. There are a number of tests that have `FindAndModify` operations (`FindAndDelete` and `FindAndReplace` in particular; here's [an example](https://github.com/mongodb/mongo-go-driver/blob/master/mongo/integration/collection_test.go#L1236)) that do not find a matching document. In that case, the `value` returned by the server will be BSON null and `famr.Value` here will be `nil`. This situation _should_ result in `mongo.ErrNoDocuments`, and now that we're correctly returning errors from `build[type]Result` functions, this `if` check needs to also accept `value`s that are BSON null.

File: x/mongo/driver/operation/hello.go [hunk 2]
```diff
@@ -1,3 +1,9 @@
+// Copyright (C) MongoDB, Inc. 2021-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
```
Comment (benjirewis) [inline x/mongo/driver/operation/hello.go:5]: `hello.go` was missing a license for some reason (the reason being I made the file and forgot to add the license).

File: x/mongo/driver/operation/find_and_modify.go (hunk 1 above)
Comment (matthewdale) [reply x/mongo/driver/operation/find_and_modify.go:79]: The behavior seems sound, but the condition below is confusing. `DocumentOK()` returns either `return nil, false` or `return doc, true`, so the condition:
```go
if famr.Value != nil && !ok
```
will never evaluate to true.

Is there something else you can check before trying to read it as a `Document`, like if the value is empty?

File: x/mongo/driver/operation/listDatabases.go [hunk 3]
```diff
@@ -62,76 +60,64 @@ func buildListDatabasesResult(response bsoncore.Document, srvr driver.Server) (L
...
-			// TODO: Make operationgen handle array results.
 			arr, ok := element.Value().ArrayOK()
 			if !ok {
-				err = fmt.Errorf("response field 'databases' is type array, but received BSON type %s", element.Value().Type)
-				continue
+				return ir, fmt.Errorf("response field 'databases' is type array, but received BSON type %s", element.Value().Type)
 			}
 
 			var tmp bsoncore.Document
 			marshalErr := bson.Unmarshal(arr, &tmp)
```
Comment (matthewdale) [inline x/mongo/driver/operation/listDatabases.go:76]: Optional: Now that we're not trying to keep track of the `err` variable value, consider renaming this variable to `err`.

Comment (benjirewis) [reply x/mongo/driver/operation/listDatabases.go:76]: Done here and elsewhere in the file.

File: x/mongo/driver/operation/command.go [hunk 4]
```diff
@@ -1,5 +1,3 @@
-// NOTE: This file is maintained by hand because operationgen cannot generate it.
-
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply x/mongo/driver/operation/command.go:2]: I think so? Added it in but cc: @kevinAlbs, do you know which files actually require licenses?

File: x/mongo/driver/operation/find_and_modify.go (hunk 1 above)
Comment (benjirewis) [reply x/mongo/driver/operation/find_and_modify.go:79]: Excellent point. I've switched the `if` check to be `if element.Value().Type != bsontype.Null && !ok`. That should correctly return an error if the `value` field is neither BSON null _nor_ a BSON document.

File: x/mongo/driver/operation/hello.go (hunk 2 above)
Comment (gabbyasuncion) [reply x/mongo/driver/operation/hello.go:5]: What's the general procedure deciding when a file should have a license? I'm not super familiar with licensing open source software but it seems pretty important to ensure it's being used the way we expect it to.

**Review by gabbyasuncion — COMMENTED** [review]
I looked at the JIRA ticket and it makes sense that we wouldn't want to ignore errors. What was the main motivation for using `operationgen` in the first place (besides code generation)? Does editing code generated by `operationgen` break it somehow?

File: x/mongo/driver/operation/find_and_modify.go (hunk 1 above)
Comment (matthewdale) [reply x/mongo/driver/operation/find_and_modify.go:79]: Makes sense 👍

**Review by matthewdale — APPROVED** [review]
Looks good 👍

**Review by gabbyasuncion — APPROVED** [review]
Looks good! Just left a few questions 👍

File: x/mongo/driver/operation/hello.go (hunk 2 above)
Comment (benjirewis) [reply x/mongo/driver/operation/hello.go:5]: A great question hahah. I'm _also_ not super familiar with our licensing policy, but this copyright notice should probably appear in every file written by us and available to users of the repo. We're certainly missing the copyright on some files ([here](https://github.com/mongodb/mongo-go-driver/blob/master/mongo/errors_test.go)'s just one example). If it's important for us to have copyright on every file, I can file a ticket for auditing the code base (@kevinAlbs).

**Pull request merged**